libmount/src/context_mount.c

```diff
@@ -776,6 +778,63 @@ static int do_mount_subdir(struct libmnt_context *cxt,
 	return rc;
 }
 
+#ifdef UL_HAVE_MOUNT_API
+/**
+ * mnt_idmapped:
+ * @target: mount to idmap
+ * @userns_fd: user namespace file descriptor to attach
+ * @recursive: whether this is a MS_REC recursive request
+ *
+ * Create an idmapped mount based on @target, unmounting the non-idmapped
+ * @target mount and attaching the detached idmapped mount @target.
+ *
+ * Returns: 0 on success,
+ *         <0 in case of error
+ */
+static int mnt_idmapped(const char *target, int userns_fd, bool recursive)
+{
+	int fd_tree = -1;
+	struct mount_attr attr = {
+		.attr_set	= MOUNT_ATTR_IDMAP,
+		.userns_fd	= userns_fd,
+
+	};
+	int rc;
+
+	/*
+	 * Once a mount has been attached to the filesystem it can't be
+	 * idmapped anymore. So create a new detached mount.
+	 */
```

Comment:
I guess, this is good enough for this patch, but later it would be probably better to rewrite do_mount() to use fsopen() and use the same mount file descriptor for fsconfig(), mount_setattr() and attache it by move_mount(). It means all the steps and additional operations will be atomic (for other userspace tasks) and without the extra umount() and open_tree(OPEN_TREE_CLONE).

I'll try to implement it during this release cycle.

Reply:
Sounds good. I'm trying to use `mount(8)` in a demo with this feature next week.

Reply:
Demo went well!